src/api-binder/report.ts

```diff
@@ -139,11 +139,26 @@ export async function startReporting() {
 	])) as StateReportOpts;
 
 	let reportPending = false;
-	const doReport = async () => {
+	let reportDeferred = false;
+	const doReport = async (): Promise<void> => {
+		// Send a report if no report is pending
 		if (!reportPending) {
 			reportPending = true;
-			await reportCurrentState(reportConfigs);
-			reportPending = false;
+			try {
+				// Report changes in the current state
+				await reportCurrentState(reportConfigs);
+			} finally {
+				reportPending = false;
+				// Check if a report was requested while this one was processing
+				if (reportDeferred) {
+					reportDeferred = false;
+					// Try to report again in case something changed while the last report was running
+					doReport();
+				}
+			}
+		} else {
+			// Defer this report so we can try again once the current one is completed
+			reportDeferred = true;
```

Comment:
You could also clear this inside `getStateAndReport()` , right below the `getCurrentForReport()`
See: https://github.com/balena-os/balena-supervisor/blob/e58683537afcc716d785c7632cb3ed3e1a22fdfa/src/api-binder/report.ts#L74

This way the deferred report is discarded, if the `getCurrentForReport()` is called after the change event, which means that the new data ended up being send already (eg b/c of a backoff).